src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchProfile.cs

```diff
@@ -1,14 +1,138 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
 
+using Microsoft.VisualStudio.Threading;
+
 namespace Microsoft.VisualStudio.ProjectSystem.Debug
 {
     /// <summary>
     /// Represents one launch profile read from the launchSettings file.
     /// </summary>
     internal class LaunchProfile : ILaunchProfile, IPersistOption
     {
-        public LaunchProfile()
+        /// <summary>
+        /// Creates a copy of <paramref name="profile"/> in which tokens are replaced via <paramref name="replaceAsync"/>.
+        /// </summary>
+        /// <remarks>
+        /// Intended to replace tokens such as environment variables and MSBuild properties.
+        /// </remarks>
+        /// <param name="profile">The source profile to copy from.</param>
+        /// <param name="replaceAsync">A function that performs token substitution.</param>
+        /// <returns>A profile with tokens substituted.</returns>
+        internal static async Task<LaunchProfile> ReplaceTokensAsync(ILaunchProfile profile, Func<string, Task<string>> replaceAsync)
+        {
+            ImmutableDictionary<string, string>? environmentVariables = profile.EnvironmentVariables;
+            ImmutableDictionary<string, object>? otherSettings = profile.OtherSettings;
+
+            if (environmentVariables != null)
+            {
+                foreach ((string key, string value) in environmentVariables)
+                {
+                    environmentVariables = environmentVariables.SetItem(key, await replaceAsync(value));
+                }
+            }
+
+            if (otherSettings != null)
+            {
+                foreach ((string key, object value) in otherSettings)
+                {
+                    if (value is string s)
+                    {
+                        otherSettings = otherSettings.SetItem(key, await replaceAsync(s));
+                    }
+                }
+            }
+
+            ImmutableArray<string> environmentVariablesKeyOrder;
+            ImmutableArray<string> otherSettingsKeyOrder;
+
+            if (profile is LaunchProfile launchProfile)
+            {
+                // Preserve the order of keys
+                environmentVariablesKeyOrder = launchProfile.EnvironmentVariablesKeyOrder;
+                otherSettingsKeyOrder = launchProfile.OtherSettingsKeyOrder;
+            }
+            else
+            {
+                // We don't know the order here, so just use dictionary order (which will be randomised by hash codes)
+                environmentVariablesKeyOrder = environmentVariables?.Keys.ToImmutableArray() ?? ImmutableArray<string>.Empty;
+                otherSettingsKeyOrder = otherSettings?.Keys.ToImmutableArray() ?? ImmutableArray<string>.Empty;
```

Comment:
Looking again, these should probably sorted alphabetically for consistency with elsewhere.